### PR TITLE
Add kubectl log --previous support to view last terminated container log

### LINKED ISSUE #4640: Feature request: add ability to view logs of pod in failed state
While working to build applications on Kubernetes, I found it difficult to view the logs of a container that went into a failed state.  I had to SSH into the node, docker ps -a, and docker logs to see the root cause.  We should provide a mechanism via kubelet to view logs of containers not yet garbage collected for access via kubectl to improve developer productivity.

/cc @smarterclayton

### PR DESCRIPTION
**Pull request by dchen1107**
Fixed #4640. 

View current running container:

```
$ cluster/kubectl.sh log kube-dns-u6iwr etcd
Starting cluster using os distro: debian
current-context: "golden-system-455_kubernetes"
Running: cluster/../cluster/gce/../../cluster/../_output/dockerized/bin/linux/amd64/kubectl log kube-dns-u6iwr etcd
2015-05-08T17:19:52.082453389Z 2015/05/08 17:19:52 etcd: no data-dir provided, using default data-dir ./default.etcd
2015-05-08T17:19:52.082507932Z 2015/05/08 17:19:52 etcd: listening for peers on http://localhost:2380
2015-05-08T17:19:52.082513303Z 2015/05/08 17:19:52 etcd: listening for peers on http://localhost:7001
2015-05-08T17:19:52.082516618Z 2015/05/08 17:19:52 etcd: listening for client requests on http://127.0.0.1:2379
2015-05-08T17:19:52.082519759Z 2015/05/08 17:19:52 etcd: listening for client requests on http://127.0.0.1:4001
2015-05-08T17:19:52.082523120Z 2015/05/08 17:19:52 etcdserver: datadir is valid for the 2.0.1 format
2015-05-08T17:19:52.082526742Z 2015/05/08 17:19:52 etcdserver: name = default
2015-05-08T17:19:52.082530076Z 2015/05/08 17:19:52 etcdserver: data dir = default.etcd
2015-05-08T17:19:52.082533163Z 2015/05/08 17:19:52 etcdserver: member dir = default.etcd/member
2015-05-08T17:19:52.082536440Z 2015/05/08 17:19:52 etcdserver: heartbeat = 100ms
2015-05-08T17:19:52.082539361Z 2015/05/08 17:19:52 etcdserver: election = 1000ms
2015-05-08T17:19:52.082542194Z 2015/05/08 17:19:52 etcdserver: snapshot count = 10000
2015-05-08T17:19:52.082545096Z 2015/05/08 17:19:52 etcdserver: advertise client URLs = http://127.0.0.1:2379,http://127.0.0.1:4001
2015-05-08T17:19:52.082553872Z 2015/05/08 17:19:52 etcdserver: initial advertise peer URLs = http://localhost:2380,http://localhost:7001
2015-05-08T17:19:52.082557335Z 2015/05/08 17:19:52 etcdserver: initial cluster = default=http://localhost:2380,default=http://localhost:7001
2015-05-08T17:19:52.082561674Z 2015/05/08 17:19:52 etcdserver: start member 6a5871dbdd12c17c in cluster f68652439e3f8f2a
2015-05-08T17:19:52.082565037Z 2015/05/08 17:19:52 raft: 6a5871dbdd12c17c became follower at term 0
2015-05-08T17:19:52.082568494Z 2015/05/08 17:19:52 raft: newRaft 6a5871dbdd12c17c [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2015-05-08T17:19:52.082571870Z 2015/05/08 17:19:52 raft: 6a5871dbdd12c17c became follower at term 1
2015-05-08T17:19:52.135010625Z 2015/05/08 17:19:52 etcdserver: added local member 6a5871dbdd12c17c [http://localhost:2380 http://localhost:7001] to cluster f68652439e3f8f2a
2015-05-08T17:19:53.778680605Z 2015/05/08 17:19:53 raft: 6a5871dbdd12c17c is starting a new election at term 1
2015-05-08T17:19:53.782947933Z 2015/05/08 17:19:53 raft: 6a5871dbdd12c17c became candidate at term 2
2015-05-08T17:19:53.782958531Z 2015/05/08 17:19:53 raft: 6a5871dbdd12c17c received vote from 6a5871dbdd12c17c at term 2
2015-05-08T17:19:53.782962231Z 2015/05/08 17:19:53 raft: 6a5871dbdd12c17c became leader at term 2
2015-05-08T17:19:53.782965857Z 2015/05/08 17:19:53 raft.node: 6a5871dbdd12c17c elected leader 6a5871dbdd12c17c at term 2
2015-05-08T17:19:53.784006311Z 2015/05/08 17:19:53 etcdserver: published {Name:default ClientURLs:[http://127.0.0.1:2379 http://127.0.0.1:4001]} to cluster f68652439e3f8f2a
```

View last terminated container:

```
$ cluster/kubectl.sh log --previous kube-dns-u6iwr etcd
Starting cluster using os distro: debian
current-context: "golden-system-455_kubernetes"
Running: cluster/../cluster/gce/../../cluster/../_output/dockerized/bin/linux/amd64/kubectl log --previous kube-dns-u6iwr etcd
2015-05-08T01:04:04.531822712Z 2015/05/08 01:04:04 etcdserver: start to snapshot (applied: 30003, lastsnap: 20002)
2015-05-08T01:04:04.539799516Z 2015/05/08 01:04:04 etcdserver: compacted log at index 30003
2015-05-08T01:04:04.567473474Z 2015/05/08 01:04:04 etcdserver: saved snapshot at index 30003
2015-05-08T02:26:45.031556653Z 2015/05/08 02:26:45 etcdserver: start to snapshot (applied: 40004, lastsnap: 30003)
2015-05-08T02:26:45.034289317Z 2015/05/08 02:26:45 etcdserver: compacted log at index 40004
2015-05-08T02:26:45.063427077Z 2015/05/08 02:26:45 etcdserver: saved snapshot at index 40004
2015-05-08T03:49:23.031419059Z 2015/05/08 03:49:23 etcdserver: start to snapshot (applied: 50005, lastsnap: 40004)
2015-05-08T03:49:23.034633854Z 2015/05/08 03:49:23 etcdserver: compacted log at index 50005
2015-05-08T03:49:23.063567111Z 2015/05/08 03:49:23 etcdserver: saved snapshot at index 50005
2015-05-08T03:49:41.240453029Z 2015/05/08 03:49:41 filePurge: successfully removed file default.etcd/member/wal/0000000000000000-0000000000000000.wal
2015-05-08T05:12:01.031479334Z 2015/05/08 05:12:01 etcdserver: start to snapshot (applied: 60006, lastsnap: 50005)
2015-05-08T05:12:01.035340614Z 2015/05/08 05:12:01 etcdserver: compacted log at index 60006
2015-05-08T05:12:01.063295479Z 2015/05/08 05:12:01 etcdserver: saved snapshot at index 60006
2015-05-08T05:12:11.283689186Z 2015/05/08 05:12:11 filePurge: successfully removed file default.etcd/member/snap/0000000000000002-0000000000002711.snap
2015-05-08T05:12:11.283939737Z 2015/05/08 05:12:11 filePurge: successfully removed file default.etcd/member/wal/0000000000000001-0000000000002712.wal
2015-05-08T06:34:41.531536560Z 2015/05/08 06:34:41 etcdserver: start to snapshot (applied: 70007, lastsnap: 60006)
2015-05-08T06:34:41.535609973Z 2015/05/08 06:34:41 etcdserver: compacted log at index 70007
2015-05-08T06:34:41.563516650Z 2015/05/08 06:34:41 etcdserver: saved snapshot at index 70007
2015-05-08T06:35:11.327789880Z 2015/05/08 06:35:11 filePurge: successfully removed file default.etcd/member/snap/0000000000000002-0000000000004e22.snap
2015-05-08T06:35:11.328059550Z 2015/05/08 06:35:11 filePurge: successfully removed file default.etcd/member/wal/0000000000000002-0000000000004e23.wal
2015-05-08T07:57:19.531724879Z 2015/05/08 07:57:19 etcdserver: start to snapshot (applied: 80008, lastsnap: 70007)
2015-05-08T07:57:19.536553856Z 2015/05/08 07:57:19 etcdserver: compacted log at index 80008
2015-05-08T07:57:19.563253209Z 2015/05/08 07:57:19 etcdserver: saved snapshot at index 80008
2015-05-08T07:57:41.373356734Z 2015/05/08 07:57:41 filePurge: successfully removed file default.etcd/member/snap/0000000000000002-0000000000007533.snap
2015-05-08T07:57:41.373606942Z 2015/05/08 07:57:41 filePurge: successfully removed file default.etcd/member/wal/0000000000000003-0000000000007534.wal
2015-05-08T09:20:00.031505826Z 2015/05/08 09:20:00 etcdserver: start to snapshot (applied: 90009, lastsnap: 80008)
2015-05-08T09:20:00.036393165Z 2015/05/08 09:20:00 etcdserver: compacted log at index 90009
2015-05-08T09:20:00.063395366Z 2015/05/08 09:20:00 etcdserver: saved snapshot at index 90009
2015-05-08T09:20:11.428310827Z 2015/05/08 09:20:11 filePurge: successfully removed file default.etcd/member/snap/0000000000000002-0000000000009c44.snap
2015-05-08T09:20:11.428828475Z 2015/05/08 09:20:11 filePurge: successfully removed file default.etcd/member/wal/0000000000000004-0000000000009c45.wal
2015-05-08T10:42:38.031654974Z 2015/05/08 10:42:38 etcdserver: start to snapshot (applied: 100010, lastsnap: 90009)
```

If the last terminated container is not exist:

```
$ cluster/kubectl.sh log --previous etcd-server-kubernetes-master
current-context: "golden-system-455_kubernetes"
Running: cluster/../cluster/gce/../../cluster/../_output/dockerized/bin/linux/amd64/kubectl log --previous etcd-server-kubernetes-master
Internal Error: previous terminated container "etcd-container" not found in pod
```

cc/ @vishh

### CODE REVIEW

File: contrib/completions/bash/kubectl [hunk 1]
```diff
@@ -404,6 +404,8 @@ _kubectl_log()
     flags+=("--help")
     flags+=("-h")
     flags+=("--interactive")
+    flags+=("--previous")
```
Comment (vishh) [inline contrib/completions/bash/kubectl]: Question: Can this be `last` instead of `previous`? And do we need the `-p` short flag?

File: docs/kubectl_log.md [hunk 2]
```diff
@@ -8,7 +8,7 @@ Print the logs for a container in a pod.
 Print the logs for a container in a pod. If the pod has only one container, the container name is optional.
 
 ```
-kubectl log [-f] POD [CONTAINER]
+kubectl log [-f] [-p] POD [CONTAINER]
```
Comment (vishh) [inline docs/kubectl_log.md]: General: We should support both `log` and `logs`.

File: docs/kubectl_log.md [hunk 3]
```diff
@@ -27,6 +30,7 @@ $ kubectl log -f 123456-7890 ruby-container
   -f, --follow=false: Specify if the logs should be streamed.
   -h, --help=false: help for log
       --interactive=true: If true, prompt the user for input when required. Default true.
+  -p, --previous=false: If true, print the logs for a previous terminated container in a pod if it exists.
```
Comment (vishh) [inline docs/kubectl_log.md]: nit: Rephrase to `print the logs of the previous instance of the container`?

**Comment by dchen1107** [comment]
--previous was chosen by our UX lead: @bgrant0607 over --last :-)

File: pkg/kubelet/dockertools/manager.go [hunk 4]
```diff
@@ -298,7 +298,7 @@ func (dm *DockerManager) inspectContainer(dockerID, containerName, tPath string)
 			if found {
 				data, err := ioutil.ReadFile(path)
 				if err != nil {
-					glog.Errorf("Error on reading termination-log %s: %v", path, err)
+					//glog.Errorf("Error on reading termination-log %s: %v", path, err)
```
Comment (vishh) [inline pkg/kubelet/dockertools/manager.go]: I assume this change is unintentional.

File: docs/kubectl_log.md (hunk 2 above)
Comment (dchen1107) [inline docs/kubectl_log.md]: what are the difference between log and logs in your view here?

File: contrib/completions/bash/kubectl (hunk 1 above)
Comment (dchen1107) [inline contrib/completions/bash/kubectl]: -p is supported already.

File: pkg/kubelet/kubelet.go [hunk 5]
```diff
@@ -1373,21 +1373,33 @@ func (kl *Kubelet) validatePodPhase(podStatus *api.PodStatus) error {
...
 	cStatus, found := api.GetContainerStatus(podStatus.ContainerStatuses, containerName)
 	if !found {
 		return "", fmt.Errorf("container %q not found in pod", containerName)
 	}
-	if cStatus.State.Waiting != nil {
-		return "", fmt.Errorf("container %q is in waiting state.", containerName)
+	if previous {
+		if cStatus.LastTerminationState.Termination == nil {
+			return "", fmt.Errorf("previous terminated container %q not found in pod", containerName)
+		} else {
```
Comment (vishh) [inline pkg/kubelet/kubelet.go]: nit: Drop `else`.

File: pkg/kubelet/kubelet.go [hunk 6]
```diff
@@ -1373,21 +1373,33 @@ func (kl *Kubelet) validatePodPhase(podStatus *api.PodStatus) error {
...
+			cID = cStatus.LastTerminationState.Termination.ContainerID
+		}
+	} else {
+		if cStatus.State.Waiting != nil {
+			return "", fmt.Errorf("container %q is in waiting state.", containerName)
+		} else {
```
Comment (vishh) [inline pkg/kubelet/kubelet.go]: Same here

File: pkg/registry/pod/rest.go [hunk 7]
```diff
@@ -212,6 +214,10 @@ func LogLocation(getter ResourceGetter, connInfo client.ConnectionInfoGetter, ct
 	if opts.Follow {
 		params.Add("follow", "true")
 	}
+	if opts.Previous {
+		params.Add("previous", "true")
+	}
+	glog.Infof("????????LogLocation=%v", opts)
```
Comment (vishh) [inline pkg/registry/pod/rest.go]: Remove debug logs

**Comment by vishh** [comment]
Just a few minor nits. Overall LGTM. Thanks for this feature @dchen1107. This will be very useful.

File: pkg/kubelet/dockertools/manager.go (hunk 4 above)
Comment (dchen1107) [inline pkg/kubelet/dockertools/manager.go]: Ahh, someone breaks that feature which generates tons of logging messages. Fixed it. Thanks!

File: docs/kubectl_log.md (hunk 2 above)
Comment (vishh) [inline docs/kubectl_log.md]: No difference. Its just convenience. For example docker's log command is `docker logs`.
This is not important for this PR.

Comment (dchen1107) [inline docs/kubectl_log.md]: Send a separate pr to handle this?

**Comment by dchen1107** [comment]
Addressed your comments, PTAL?

File: docs/kubectl_log.md (hunk 2 above)
Comment (vishh) [inline docs/kubectl_log.md]: Yup!

**Comment by vishh** [comment]
LGTM. WIll merge once tests pass.

File: docs/kubectl_log.md (hunk 2 above)
Comment (derekwaynecarr) [inline docs/kubectl_log.md]: +1 to support logs and log

**Comment by derekwaynecarr** [comment]
thanks for adding this and sorry I was not able to review it earlier.  It lgtm and and is a big help when actually building apps on Kube
